packages/lesswrong/server/resolvers/ultraFeedSubscriptionsResolver.ts

```diff
@@ -0,0 +1,354 @@
+      posts: allPostIds,
+      comments: commentIds
+    });
+
+    // Get read statuses for all posts that appear in comment threads
+    const postReadStatuses = await postsRepo.getPostReadStatuses(postIdsFromComments, currentUser?._id ?? null);
+
+    const targetsByHourAndThread = new Map<string, Map<string, SliceTarget[]>>();
+
+    for (const c of subscriptionComments) {
+      if (!c.isInitialCandidate || !c.postedAt) continue;
+      const hourKey = new Date(c.postedAt);
+      hourKey.setMinutes(0, 0, 0);
+      const hourStr = hourKey.toISOString();
+      const threadId = c.topLevelCommentId ?? c.commentId;
+      if (!targetsByHourAndThread.has(hourStr)) {
+        targetsByHourAndThread.set(hourStr, new Map());
+      }
+      const byThread = targetsByHourAndThread.get(hourStr)!;
+      if (!byThread.has(threadId)) {
+        byThread.set(threadId, []);
+      }
+      byThread.get(threadId)!.push({ commentId: c.commentId, postedAt: c.postedAt });
+    }
+
+    const commentsByThread = new Map<string, typeof subscriptionComments>();
+    for (const c of subscriptionComments) {
+      const threadId = c.topLevelCommentId ?? c.commentId;
+      if (!commentsByThread.has(threadId)) {
+        commentsByThread.set(threadId, []);
+      }
+      commentsByThread.get(threadId)!.push(c);
+    }
+
+    type FeedItem = { 
+      type: SubscribedFeedEntryType; 
+      sortDate: Date; 
+      data: {
+        _id: string;
+        post?: DbPost | null;
+        postMetaInfo?: FeedPostMetaInfo;
+        comments?: DbComment[];
+        commentMetaInfos?: Record<string, FeedCommentMetaInfo>;
+        isOnReadPost?: boolean;
+        postSources?: string[];
+        users?: DbUser[];
+      };
+    };
+    const feedItems: FeedItem[] = [];
+
+    for (const row of postRows) {
+      const post = row.post;
+      if (!post?._id) continue;
+      if (hideRead && (row.postMetaInfo?.lastViewed || row.postMetaInfo?.lastInteracted)) {
+        continue;
+      }
+      feedItems.push({
+        type: 'feedPost',
+        sortDate: post.postedAt ?? new Date(),
+        data: {
+          _id: post._id,
+          post: post as DbPost,
+          postMetaInfo: { 
+            servedEventId: randomId(), 
+            sources: ['subscriptionsPosts'] as const,
+            lastViewed: row.postMetaInfo?.lastViewed ?? null,
+            lastInteracted: row.postMetaInfo?.lastInteracted ?? null,
+            highlight: !(row.postMetaInfo?.lastViewed || row.postMetaInfo?.lastInteracted),
+            displayStatus: 'expanded' as const,
+            isRead: !!(row.postMetaInfo?.lastViewed || row.postMetaInfo?.lastInteracted),
+          },
+        },
+      });
+    }
+
+    function buildPathsForThread(threadId: string) {
+      const commentsInThread = commentsByThread.get(threadId) ?? [];
+      return buildDistinctLinearThreads(commentsInThread);
+    }
+
+    for (const [hourStr, byThread] of targetsByHourAndThread) {
```

Comment:
The rest of this resolver is going to be pretty difficult to review given I don't know what the intended logic is and it's mostly undocumented.  Maybe it doesn't matter?